경북대 BE_유우석 Step2 - 회원 로그인

milk-stone opened the pull request:
Step2 기능 구현하여 제출합니다!

이번 과제에서 생각보다 애를 많이 먹었던 것 같습니다.

### 1. 사용자 테이블 이름을 user로 하여 발생했던 문제

- USER는 SQL의 예약어(Reserved Keyword)이기 때문에, 테이블 이름으로 사용할 수 없습니다.
```
CREATE TABLE user (

id BIGINT AUTO_INCREMENT PRIMARY KEY,

email VARCHAR(255) NOT NULL,

password VARCHAR(255) NOT NULL,

name VARCHAR(255) NOT NULL,

role VARCHAR(10) NOT NULL

);
```
따라서, 위와 같은 CREATE TABLE 문을 사용할 수 없습니다.

그렇기에 사용자 정보를 저장하는 TABLE 이름을 Member로 수정하였는데, 이를커밋 내용에 적지 못했습니다. ㅠㅠ

### 2. Postman 자체 오류

<img width="431" alt="image" src="https://github.com/user-attachments/assets/a3fc14d4-7f08-4112-ac3b-6ceeefb4445c" />

이렇게 외관상 API 엔드포인트가 잘 적혀있는 것 같지만,

```
2025-07-08T00:41:27.019-07:00 ERROR 66184 --- [spring-gift] [nio-8080-exec-1] g.g.exception.GlobalExceptionHandler     : 서버 내부에서 처리되지 않은 예외가 발생했습니다.



org.springframework.web.servlet.resource.NoResourceFoundException: No static resource api.

at org.springframework.web.servlet.resource.ResourceHttpRequestHandler.handleRequest(ResourceHttpRequestHandler.java:585) ~[spring-webmvc-6.2.8.jar:6.2.8]

at..
```
이런 오류가 발생했습니다.

알고보니 포스트맨 애플리케이션 오류로, 실제로 엔드포인트가 `/api` 까지만 적혀서 날아가고 있었습니다.

포스트맨을 껐다 켜서 해결했습니다.

milk-stone committed:
refactor: 숫자 가독성을 위한 _ 추가

milk-stone committed:
docs: Step 2 에서 구현할 기능 목록 작성

milk-stone committed:
feat: JWT 생성을 위한 의존성 추가

milk-stone committed:
feat: User table이 추가됨에 따라, user table에 대한 CREATE문, INSERT문 추가

milk-stone committed:
feat: 사용자 설정 예외 처리를 하기 위한 클래스 추가 및 폴더 구조 변경

milk-stone committed:
style: 폴더 구조 변경

gift.product 에서, gift.domain.product 로 사이에 domain, global 폴더 구조 추가

milk-stone committed:
feat: 회원 가입 API, 로그인 API 작성

1. 회원 가입 기능
- 사용자는 이메일, 비밀번호, 이름, 역할을 입력하면, 해당 내용을 가지는 Member 객체를 저장하고, AccessToken을 발행한다.

** 왜 요구사항에 없는 이름, 역할 필드를 추가하였는지에 대한 서술 **
- Token을 발행할 때, generateToken 함수를 사용하는데, 여기에 .claim 함수는 사용자 정보 외에도 추가로 토큰에 담을 정보를 설정하는 함수입니다. 따라서 당연히 이 name 필드와 role 필드가 없어도 토큰을 생성할 수 있습니다. 그러나, 저는 .claim 함수를 사용하고 싶었고 그렇기에 member 테이블에 name, role 필드를 추가하였습니다.

2. 로그인 기능
- 사용자는 이메일과 패스워드를 입력하면, 사용자의 이메일에 대한 비밀번호가 일치한지 확인한 후 일치하는 경우에 AccessToken을 발행한다.

milk-stone committed:
feat: Member 엔티티 구성

name 필드와 role 필드를 추가하였음.

그에 따라, schema.sql과 data.sql를 수정하였음.

milk-stone committed:
style: 폴더 구조 변경

milk-stone committed:
style: 사용자 CRUD 기능을 작성하기 위한 뼈대 코드 작성

milk-stone committed:
docs: Step 구현 내용 체크

Rok93 commented on src/main/java/gift/domain/member/controller/MemberController.java:
```diff
@@ -0,0 +1,15 @@
+
+@RestController
+@RequestMapping("/api/members")
+public class MemberController {
```
이 컨트롤러는 아무런 API도 없는 것 같은데, 이 부분은 혹시 커밋이 누락된걸까요? 🤔
혹시 사용하지 않는 컨트롤러라면 제거 부탁드려요!

milk-stone replied:
유저 관리자 페이지를 만들려고 하다가 구현을 못해서 뼈대만 작성한 상황이었습니다! 삭제했다가 추후에 관리자 페이지 만들 때 다시 재작성하겠습니다!

Rok93 commented on src/main/java/gift/domain/member/controller/AuthController.java:
```diff
@@ -0,0 +1,32 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/auth")
```
[API 명세](https://edu.nextstep.camp/s/UH6H8Iba/ls/TxIElkRG)에서 이미 스펙이 정해진 API들로 보이는데, 명세에 맞게 엔드포인트 수정 부탁드릴게요. 
같이 협업하는 클라이언트 개발자(= 프론트엔드, iOS, 안드로이드 etc..)는 API 명세에 맞게 작업을 하고있을텐데 API 응답을 제공하는 백엔드 개발자가 전혀 다른 스펙으로 작업을하게된다면 클라이언트개발자는 이중으로 작업을 해야할거에요. 🥲 

<img width="784" alt="image" src="https://github.com/user-attachments/assets/a80327c7-0e78-41c7-bb08-2e76c8c400d1" />

milk-stone replied:
아. 바로 고치겠습니다! 정신이 없었네요,, 정확하게 수정해서 재 제출 하겠습니다!

Rok93 commented on src/main/java/gift/domain/member/dto/LoginRequest.java:
```diff
@@ -0,0 +1,4 @@
+package gift.domain.member.dto;
+
+public record LoginRequest(String email, String password) {
```
1단계에서 했던 것처럼 email, passowrd 요청에 대해서도 기본적인 유효성 검증은 추가해주면 어떨까요?  (다른 request 객체는 전체적으로 확인 부탁드릴게요 🙏)

milk-stone replied:
넵!! Not Null 과 같은 기본 유효성 검증 로직을 추가하겠습니다!

Rok93 commented on src/main/java/gift/domain/member/service/AuthService.java:
```diff
@@ -0,0 +1,96 @@
+        this.memberRepository = memberRepository;
+    }
+
+    String secretKey = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
```
secretKey의 접근제한자를 default로 두셨는데, 이유가 있으신가요? 🤔
secretKey도 memberRepository와 동일하게 AuthService의 필드로 둘 수 있지 않을까요?

milk-stone replied:
아이고.. 제가 다시 봐도 지저분한 것들이 너무 많네요.. 이런 코드 보여드려서 죄송한 마음입니다.. 수정해서 최대한 깔끔하게 작성해보겠습니다!

Rok93 commented on src/main/java/gift/domain/member/service/AuthService.java:
```diff
@@ -0,0 +1,96 @@
+        this.memberRepository = memberRepository;
+    }
+
+    String secretKey = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
```
추가적으로 이런 [설정 값들은 properties에 두고 주입](https://www.baeldung.com/properties-with-spring)받아보면 어떨까요? 

이 방식의 장점은 무엇일까요? 😃

milk-stone replied:
외부에 노출이 되지 않기 때문에 좋을 것이라고 생각합니다!

만약 properties에 작성한 후 주입받아도, application.properties 코드가 깃허브 상에 공개된다면, 이 또한 노출이 되는 것 같은데 properties에 두고 주입받으면 더 좋은 점이 어떤게 있는지 여쭤보고 싶습니다!

추후 구현 방법: secretKey를 환경변수로 빼두고, properties에 설정해두어 주입을 받되, properties 파일에는 환경변수만 설정해놓는 방법으로 구현할 것 같습니다!

Rok93 replied:
환경 변수도 방안이될 수 있지만, 환경 변수가 너~~무 많아지는 경우에는 이 방식이 쉽지는 않을 것 같다는 생각도 드네요.  🤔 

말씀하신 것처럼 결국 properties 파일도 레포지토리에 커밋&푸시를 하게될텐데 만약 레포지토리가 private이라면 그나마 괜찮겠지만 public 레포지토리라면 외부에 이런 비밀키가 공개되기 떄문에 보안에 문제가 생기게 됩니다. 

그렇기 때문에 운영환경의 properties는 레포지토리에 올리지 않고 배포 시점에 가져와 주입한다거나? 혹은 key와 같이 보안성이 필요한 것들은 AWS의 SecreteManager와 같이 공개하지 않도록하는 보안책이 필요하게되구요. 😃 

저는 일부 방법만 말씀드렸는데, 이 외에도 저도 모르는 다양한 방법들이 있을 것 같아요. 
상황에따라 적절한 방식을 골라서 사용하면될 것 같아요.

Rok93 commented on src/main/java/gift/domain/member/service/AuthService.java:
```diff
@@ -0,0 +1,96 @@
+
+        Optional<Member> optionalMember = memberRepository.findByEmail(email);
+        if (optionalMember.isPresent()){
+            throw new MemberNotFoundException("AuthService : signIn() failed - Already registered member");
```
멤버가 존재할 때,  멤버 존재하지 않음 예외를 던지는 것은 조금 이상하지 않나요? 🤔
Custom Exception을 쓰시기로하셨다면 상황에 맞는 적절한 Cusom Exception을 추가해주심이 좋을 것 같아요.

milk-stone replied:
BadRequestException으로 작성하려고 생각했었는데 이렇게 작성했었네요,, 더 신중하게 작성해보겠습니다.

궁금한 것이, 이렇게 Custon Exception을 작성하면서 느끼는건데 이렇게 Exception Class가 많아지는게 좋은 방법일까요..?? 혹시 더 좋은 예외 처리 패턴이 있을까요?

Rok93 replied:
이와 관련해서는 [좋은 글](https://tecoble.techcourse.co.kr/post/2020-08-17-custom-exception/)이 있어서 대답을 대신해보려고합니다. 😃 

결국 장/단점이 있기 때문에 상황에 맞는 적절한 방식을 택하는 것이 좋다고 생각해요.

Rok93 commented on src/main/java/gift/domain/member/service/AuthService.java:
```diff
@@ -0,0 +1,96 @@
+        String email = signInRequest.email();
+        String password = signInRequest.password();
+
+        Optional<Member> optionalMember = memberRepository.findByEmail(email);
+        if (optionalMember.isPresent()){
```
memberRepository 객체에게 단순히 `email의 유저가 존재하는지 여부`만 질문할 수는 없을까요? 
결과 값으로는 true, false 를 반환받는다면 더 깔끔한 구현이 될 수 있을 것 같아요.

milk-stone replied:
매 번 이런 통찰을 주셔서 감사합니다! 훨씬 코드가 깔끔해질 것 같습니다! 반영하겠습니다.

Rok93 commented on src/main/java/gift/domain/member/service/AuthService.java:
```diff
@@ -0,0 +1,96 @@
+        if (optionalMember.isEmpty()){
+            throw new MemberNotFoundException("AuthService : login() failed - 404 Not Found Error");
+        }
+        Member user = optionalMember.get();
```
[Java Optional 바르게 쓰기](https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/) 글을 한번 읽어보신 뒤, 지금 작성하신 코드를 개선해보시면 좋을 것 같습니다. 😃

milk-stone replied:
한 번 읽어봤는데, 제가 지키지 않고 있는 부분이 정말 많네요. 다른 수정사항 다 반영한 후 이전에 작성했던 코드에 대해서도 Optional 바르게 쓰기 규칙을 지키도록 수정해보겠습니다!

Rok93 commented on src/main/java/gift/domain/member/service/AuthService.java:
```diff
@@ -0,0 +1,96 @@
+            throw new MemberNotFoundException("AuthService : login() failed - 404 Not Found Error");
+        }
+        Member user = optionalMember.get();
+        if (!user.getPassword().equals(password)){
```
user 객체에게 [메시지를 보내보면](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/) 어떨까요?

milk-stone replied:
그러면 더욱 책임 분리가 잘 될 것 같은 생각이 드네요! 수정하겠습니다.

Rok93 commented on src/main/java/gift/domain/member/service/AuthService.java:
```diff
@@ -0,0 +1,96 @@
+        return new TokenResponse(accessToken);
+    }
+
+    private String generateToken(String email, String password, Member member){
+        return Jwts.builder()
+                .setSubject(member.getId().toString())
+                .claim("name", member.getName())
+                .claim("role", member.getRole())
+                .signWith(Keys.hmacShaKeyFor(secretKey.getBytes()))
+                .compact();
```
현재 JWT 토큰을 생성하거나 혹은 읽는 역할을 AuthService에서 수행하고있는데요. 

이 역할은 별도의 객체에게 위임해보면 어떨까요?

milk-stone replied:
JwtProvider 서비스 클래스를 만들어 위임해보겠습니다!

Rok93 commented on src/main/java/gift/domain/member/service/AuthService.java:
```diff
@@ -0,0 +1,96 @@
+        return new TokenResponse(accessToken);
+    }
+
+    private String generateToken(String email, String password, Member member){
```
email이나 password는 토큰을 생성할 때, 사용하지 않는 것 같은데, 사용하지 않는 파라미터를 굳이 메서드로 받을 필요가 없지 않을까요? 🤔

milk-stone replied:
수정했습니다! email, password 정보도 넣으려고 했어서 파라미터로 넣었었는데, 마지막에 안 넣는 방향으로 생각을 바꿨지만 코드 수정은 안했었네요ㅠㅠ

Rok93 commented on src/main/java/gift/domain/member/service/AuthService.java:
```diff
@@ -0,0 +1,96 @@
+                .compact();
+    }
+
+    public String extractEmailFromAccessToken(String accessToken) throws TokenExpiredException {
```
마찬가지로 이 메서드도 사용하지 않는 것으로 보이네요. 
사용하지 않는다면 과감히 제거해주세요. 🙏

milk-stone replied:
알겠습니다! 토큰 인증을 위해서 사용할 예정이라 이 메서드를 사용하는 코드를 작성해서 리뷰요청 하겠습니다!

Rok93 commented on src/main/java/gift/global/exception/GlobalExceptionHandler.java:
```diff
@@ -38,4 +38,31 @@ public ResponseEntity<SingleErrorResponse> handleExceptions(Exception e) {
+        SingleErrorResponse errorResponse = new SingleErrorResponse(e.getMessage());
+        return new ResponseEntity<>(errorResponse, HttpStatus.NOT_FOUND);
+    }
+    @ExceptionHandler(LoginFailedException.class)
+    public ResponseEntity<SingleErrorResponse> handleLoginFailedException(LoginFailedException e) {
+        log.warn("로그인 실패 : {}", e.getMessage());
+        SingleErrorResponse errorResponse = new SingleErrorResponse(e.getMessage());
+        return new ResponseEntity<>(errorResponse, HttpStatus.FORBIDDEN);
+    }
```
로그인 실패의 경우 403응답을 내려주는게 맞을까요? 
[Http Status Code](https://developer.mozilla.org/ko/docs/Web/HTTP/Reference/Status)를 살펴보신 뒤, 더 적절한 상태 코드가 없는지 고민해보시면 좋을 것 같아요. 😃

milk-stone replied:
<img width="927" alt="image" src="https://github.com/user-attachments/assets/365af3a5-6e53-430c-94b3-4231575f59b8" />

문제 페이지에 403 forbidden을 반환하라고 되어있어서 이렇게 구현했는데 더 적절한 상태 코드가 있을지 고민해보겠습니다!

Rok93 replied:
@milk-stone 
401 응답과 403 응답은 각각 무엇이고 어떤 경우에 사용해야할까요? 
이 질문에 먼저 우석님의 생각을 남겨주시면 좋을 것 같아요!  😃

[Http Status Code](https://developer.mozilla.org/ko/docs/Web/HTTP/Reference/Status)

milk-stone replied:
401 Unauthorized : 서버가 사용자가 누군지 전혀 알지 못하는 상황에서 발생하는 에러
403 Forbidden : 서버는 사용자가 누군지 알지만(인증은 됐으나) 권한이 부족할 때 발생하는 에러

로그인 실패의 경우 서버가 사용자를 전혀 알지 못하는 상황이네요! 401 에러가 적절한 것 같습니다!

Rok93 commented on README.md:
```diff
@@ -12,4 +12,46 @@
+
+## Step 2. 회원 로그인
+
+사용자가 회원 가입, 로그인, 추후 회원별 기능을 이용할 수 있도록 구현한다.
```
[2단계 미션 요구사항](https://edu.nextstep.camp/s/UH6H8Iba/ls/oULf3oxF)의 힌트를 살펴보시면 `인증` 방식에 대한 힌트도 설명하고있는데요. 이번 미션은 단순히 JWT 토큰을 발급하는 것에서 끝나는 것이 아니라 발급된 JWT를 기준으로 인증을 수행하는 부분도 진행을 해주시는 것이 요구사항이었다고 생각해요.  

아마 이와 관련해서는 기능 구현을 안해주신 것 같아서 다음 리뷰 요청 때는 해당 기능도 추가해주세요. 🙏

milk-stone replied:
넵 알겠습니다!

Rok93 reviewed (changes requested):
우석님 2단계도 잘 진행해주셨습니다. 👍
몇몇 코멘트 남겨두었는데 확인해서 반영 부탁드릴게요. 🙏

milk-stone committed:
chore: secretKey 관련, 키 값을 환경변수로 분리하고, @Value로 값을 주입하는 형식으로 변경

dev.env 파일에 환경변수 저장, application.properties에 jwt.secret 설정

milk-stone committed:
refactor: secretKey 값을 디폴트 필드로 정의했던 것을 @Value로 주입받도록 수정

milk-stone committed:
refactor: Jwt 토큰 관련 작업을 수행하는 JwtProvider 클래스 작성

milk-stone committed:
fix: email 필드가 unique 특성을 가지도록 추가

milk-stone committed:
refactor: 3가지 변경 사항 적용

1. email 존재 여부 확인하는 메서드 작성 후 이용하도록 수정

2. Optional 바르게 쓰기 규칙 지키도록 수정

3. Member 객체에게 message를 보내서 비밀번호가 같은지 확인하도록 수정

milk-stone committed:
fix: 현재 로그인, 회원 가입 API 엔드포인트를 NEXTSTEP API 명세서 엔드포인트와 동일하게 수정

milk-stone committed:
feat(dto): 입력 유효성 검증 코드 추가

email, password, name, role 필드에 notblank 검증 적용

milk-stone committed:
feat: Token으로 Member 객체를 받아오는 메서드 작성

milk-stone committed:
style: auth 패키지 생성 후 파일 경로 변경

milk-stone committed:
feat: CustomException class 2개 추가 후  GlobalExceptionHandler 에 설정

milk-stone committed:
feat: Member의 role를 USER, ADMIN 으로 제한하기 위해 enum 작성 및 Member Entity 수정

milk-stone committed:
feat: role를 회원 가입, 유저 정보 수정 시 수정하지 못하도록 필드 삭제 (member 생성 시 USER로 자동 생성됨)

milk-stone committed:
feat: JwtProvider 수정

1. SecretKey를 import하여 비밀키 설정
2. 토큰 만료 기한을 설정 (3600000 ms = 1 hours)
3. 토큰에 저장하는 정보 셋 변경 및 추가
  - access token임을 명시 (추후 refreshToken을 추가할 경우 대비)
  - member id 대신 member email 정보 저장
  - 현재 시각 정보 추가
  - 만료 기한 정보 추가
  - TokenExpiredException 이 extends Exception을 하여 Throws를 명시해줬어야 했는데, extends RuntimeException 으로 수정하여 try catch, throws 삭제

milk-stone committed:
fix: role이 ADMIN USER로 제한됨에 따라 초기에 저장하는 데이터 셋의 role column 내용 변경

milk-stone committed:
refactor : 멤버 존재 여부를 email이 존재하는지 체크하는 메서드를 만들어 사용, Bearer 토큰을 받기에, 앞에 Bearer 를 제거하는 로직 추가, 비밀번호 확인시 user에 메세지를 보내도록 수정

milk-stone committed:
feat: 만료 기한을 application.properties에 명시

milk-stone committed:
feat: 관리자 페이지에서 사용할 유저 정보 열람, 수정, 삭제 API 작성

milk-stone committed:
style : debug를 위해 사용했던 sout 코드 삭제

Rok93 reviewed (changes requested):
우석님 현재 애플리케이션 실행이 안되는 것 같은데, 확인한번 부탁드릴게요. 🙏 
리뷰요청 전에는 테스트 코드나 애플리케이션이 잘 실행되는지 꼭 한번 확인 부탁드릴게요. 🤗

milk-stone committed:
chore: 환경 변수로 빼두었던 secret key 값 application.properties에 작성

milk-stone commented:
> 우석님 현재 애플리케이션 실행이 안되는 것 같은데, 확인한번 부탁드릴게요. 🙏 리뷰요청 전에는 테스트 코드나 애플리케이션이 잘 실행되는지 꼭 한번 확인 부탁드릴게요. 🤗

jwt.secret key 값을 환경변수로 빼두고 gitignore 처리했었다는걸 뒤늦게 알게됐네요.. 이 방식이 멘토님이 리뷰하기에 좋지 않은 방식이라는 걸 깨달았습니다! 수정했습니다!

Rok93 reviewed (approved):
우석님 피드백 반영 잘해주셨네요. 👍👍 
2단계는 이만 머지하도록할게요. 🎉 
3단계로 넘어가보시죠! 🚀🚀 

추가로 몇몇 코멘트 남겨두었는데 3단계 진행전에 확인 부탁드릴게요. 🤗
